### PR TITLE
Migrate to Rust 2018

### PR DESCRIPTION
**Pull request by Centril**
r? @eddyb cc @qmx

### CODE REVIEW

File: macros/tests/json.rs [hunk 1]
```diff
@@ -1,6 +1,5 @@
 extern crate gll;
 extern crate gll_macros;
```
Comment (eddyb) [inline macros/tests/json.rs]: Are these really needed?

File: src/generate/rust.rs [hunk 2]
```diff
@@ -16,14 +16,14 @@ pub trait RustInputPat {
     fn rust_matcher(&self) -> Src;
 }
 
-impl<S: AsRef<str>> RustInputPat for ::scannerless::Pat<S> {
+impl<S: AsRef<str>> RustInputPat for crate::scannerless::Pat<S> {
```
Comment (eddyb) [inline src/generate/rust.rs]: At this point, I'd say just import the `scannerless` module.

File: macros/tests/json.rs (hunk 1 above)
Comment (Centril) [reply macros/tests/json.rs]: I couldn't get rid of it :( Perhaps try to remove them after this PR since you're more familiar?

File: src/generate/rust.rs (hunk 2 above)
Comment (Centril) [reply src/generate/rust.rs]: Let's do that in a follow up?

File: src/generate/rust.rs [hunk 3]
```diff
@@ -64,7 +64,7 @@ impl<Pat: PartialEq> RuleWithNamedFields<Pat> {
                 .collect();
             for (field, paths) in &self.fields {
                 for path in paths {
-                    if path.len() == 0 {
```
Comment (eddyb) [inline src/generate/rust.rs]: I think this should be using slice pattern-matching on `path` (if that's a thing on stable), instead of checking the length and using `path[0]`.
The patterns would be `[]`, `[variant]` and `[variant, ..]`, I think?

File: src/generate/rust.rs [hunk 4]
```diff
@@ -136,13 +136,13 @@ impl<Pat> Rule<Pat> {
 // FIXME(eddyb) this should just work with `self: &Rc<Self>` on inherent methods,
 // but that still requires `#![feature(arbitrary_self_types)]`.
 trait RcRuleRuleMapMethods<Pat>: Sized {
-    fn parse_node_kind(&self, rules: &RuleMap<Pat>) -> ParseNodeKind;
-    fn parse_node_desc(&self, rules: &RuleMap<Pat>) -> String;
-    fn fill_parse_node_shape(&self, rules: &RuleMap<Pat>);
+    fn parse_node_kind(&self, rules: &RuleMap<'_, Pat>) -> ParseNodeKind;
+    fn parse_node_desc(&self, rules: &RuleMap<'_, Pat>) -> String;
+    fn fill_parse_node_shape(&self, rules: &RuleMap<'_, Pat>);
```
Comment (eddyb) [inline src/generate/rust.rs]: Ugh, this silly thing again. I wish we decided on the clean solution...

File: src/generate/rust.rs [hunk 5]
```diff
@@ -512,11 +505,9 @@ macro_rules! thunk {
...
         let prefix = quote!($($t)*);
         Thunk::new(move |mut cont| {
-            // HACK(eddyb) remove awkward scope post-NLL
-            {
-                let code = cont.to_inline();
-                *code += mem::replace(code, prefix);
-            }
+            let code = cont.to_inline();
+            let old_code = mem::replace(code, prefix);
+            *code += old_code;
```
Comment (eddyb) [inline src/generate/rust.rs]: Does `*code += mem::replace(code, prefix)` not work?

File: src/generate/rust.rs [hunk 6]
```diff
@@ -858,24 +838,24 @@ where
     let rust_slice_ty = Pat::rust_slice_ty();
     quote!(
         impl<'_a, I> #ident<'_a, 'static, I>
-            where I: ::gll::runtime::Input<Slice = #rust_slice_ty>,
+            where I: crate::gll::runtime::Input<Slice = #rust_slice_ty>,
```
Comment (eddyb) [inline src/generate/rust.rs]: Oh noes, this is baaad. Can we do the PR in two steps? First, update the generator, and then the generated code?
Or at least I'm curious what errors we'd get, if we change the edition but add `#[allow(rust_2018_idioms)]` on `mod parse_grammar;`.

File: src/generate/rust.rs (hunk 5 above)
Comment (Centril) [reply src/generate/rust.rs]:
```
error[E0499]: cannot borrow `*code` as mutable more than once at a time
   --> src/generate/rust.rs:509:35
    |
509 |             *code += mem::replace(code, prefix);
    |             ----------------------^^^^---------
    |             |                     |
    |             |                     second mutable borrow occurs here
    |             first mutable borrow occurs here
    |             first borrow later used here
...
771 |                     + opt(thunk!(assert_eq!(_range.start(), c.fn_input.start());)
    |                           ------------------------------------------------------- in this macro invocation
```

File: src/generate/rust.rs [hunk 7]
```diff
@@ -944,7 +924,7 @@ where
         };
         quote!(
             #[allow(non_camel_case_types)]
-            pub struct #ident<'a, 'i: 'a, I: 'a + ::gll::runtime::Input> {
+            pub struct #ident<'a, 'i: 'a, I: crate::gll::runtime::Input> {
```
Comment (eddyb) [inline src/generate/rust.rs]: We can get rid of `'i: 'a` too, right? Can we do some of these in a separate PR, before the move to 2018?

File: src/generate/templates/header.rs [hunk 8]
```diff
@@ -143,8 +143,8 @@ pub enum ListHead<C> {
     Nil,
 }
 
-impl<'a, 'i, I: ::gll::runtime::Input, T> Handle<'a, 'i, I, [T]> {
-    fn one_list_head(self) -> ListHead<Result<(Handle<'a, 'i, I, T>, Handle<'a, 'i, I, [T]>), Ambiguity<Self>>> {
+impl<'a, 'i, I: crate::gll::runtime::Input, T> Handle<'a, 'i, I, [T]> {
+    fn one_list_head(self) -> ListHead<Result<(Handle<'a, 'i, I, T>, Self), Ambiguity<Self>>> {
```
Comment (eddyb) [inline src/generate/templates/header.rs]: Heh, there's a similar one... somewhere. I think in an `-> impl Iterator`? Which used to be broken?
Maybe we changed that one and not this one.

File: src/grammar.rs [hunk 9]
```diff
@@ -242,35 +242,27 @@ impl<Pat> BitOr for RuleWithNamedFields<Pat> {
...
             _ => (
                 &[][..],
-                // HACK(eddyb) replace with `Some(self)` post-NLL
-                Some(Self {
-                    rule: self.rule.clone(),
-                    fields: self.fields,
-                }),
+                Some(self),
                 OrderMap::new(),
             ),
```
Comment (eddyb) [inline src/grammar.rs]: Does this not get reformatted onto one line?

File: src/high.rs [hunk 10]
```diff
@@ -44,7 +44,7 @@ type_lambda! {
 // FIXME(#52812) replace with `&'a <T as ApplyL<'b>>::Out`
 pub struct RefApplyL<'a, 'b, T: LambdaL>(&'a <T as ApplyL<'b>>::Out);
 
-impl<'a, 'b, T: LambdaL> Deref for RefApplyL<'a, 'b, T> {
+impl<'b, T: LambdaL> Deref for RefApplyL<'_, 'b, T> {
```
Comment (eddyb) [inline src/high.rs]: I think you're fine if you rename `'b` to `'a`.

File: src/generate/rust.rs (hunk 7 above)
Comment (Centril) [reply src/generate/rust.rs]: Yeah that worked.

File: macros/tests/json.rs (hunk 1 above)
Comment (eddyb) [reply macros/tests/json.rs]: I found the reason, it's elsewhere.

**Comment by eddyb** [comment]
So I don't repeat myself everywhere: I'd rather minimize this PR's diff, so potentially do cleanup *before*.
I don't want unnecessary changes in a 2018-ification, because it may be harder to figure out what happened, *afterwards*.

File: src/generate/rust.rs (hunk 5 above)
Comment (eddyb) [reply src/generate/rust.rs]: That's... suboptimal, ugh. I guess you can rename `old_code` to `suffix`.

File: src/generate/rust.rs [hunk 11]
```diff
@@ -64,20 +64,18 @@ impl<Pat: PartialEq> RuleWithNamedFields<Pat> {
                 .collect();
             for (field, paths) in &self.fields {
                 for path in paths {
-                    if path.is_empty() {
-                        return None;
-                    }
-                    if path.len() == 1 {
-                        if variants[path[0]].name != "" {
-                            return None;
+                    match &path[..] {
```
Comment (eddyb) [inline src/generate/rust.rs]: I think you can just use `path[..]`?

File: src/generate/rust.rs [hunk 12]
```diff
@@ -64,20 +64,18 @@ impl<Pat: PartialEq> RuleWithNamedFields<Pat> {
...
+                        [] => return None,
+                        [path] if variants[*path].name != "" => return None,
+                        [path] => variants[*path].name = field,
```
Comment (eddyb) [inline src/generate/rust.rs]: `variant`, not `path`.

File: src/generate/rust.rs [hunk 13]
```diff
@@ -64,20 +64,18 @@ impl<Pat: PartialEq> RuleWithNamedFields<Pat> {
...
+                        // FIXME: use [path_0, rest @ ..] when possible.
```
Comment (eddyb) [inline src/generate/rust.rs]: `variant`, not `path_0`.
